### PR TITLE
[Pytorch] code for 17.1: GAN FIX #1300

### LINKED ISSUE #1300: [pytorch] code for 17.1: GAN
I'm trying to translate [GAN](http://preview.d2l.ai/d2l-en/master/chapter_generative-adversarial-networks/gan.html) to pytorch.

When I search, then I found someone has translated it in https://www.pythonf.cn/read/33854.
But I have some questions about the Sigmoid layer.

image

http://preview.d2l.ai/d2l-en/master/chapter_generative-adversarial-networks/gan.html

I have tried to read https://mxnet.incubator.apache.org/versions/1.6/api/python/docs/api/gluon/nn/index.html#mxnet.gluon.nn.Activation.

There is no source code. What is the default Activation? Do I need to add the Sigmoid layer?

### PR DESCRIPTION
**Pull request by StevenJokes**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by StevenJokes** [comment]
@AnirudhDagar

**Comment by mli** [comment]
Job d2l-en/PR-1308/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1308/

**Comment by StevenJokes** [comment]
I'm tired to merge PR  too slowly. And I don't have a job now, maybe I should quit.

File: chapter_generative-adversarial-networks/gan.md [hunk 1]
```diff
@@ -42,65 +42,112 @@ $$min_D max_G \{ -E_{x \sim \text{Data}} log D(\mathbf x) - E_{z \sim \text{Nois
...
 npx.set_np()
 ```
 
+```{.python .input}
+#@tab pytorch
+%matplotlib inline
+from d2l import torch as d2l
+import torch
+from torch.utils.data import DataLoader
+from torch.autograd import Variable
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/gan.md:59]: Please don't make use of Variable.

Comment (AnirudhDagar) [reply chapter_generative-adversarial-networks/gan.md:59]: Remove the import

File: chapter_generative-adversarial-networks/gan.md [hunk 2]
```diff
@@ -42,65 +42,112 @@ $$min_D max_G \{ -E_{x \sim \text{Data}} log D(\mathbf x) - E_{z \sim \text{Nois
...
-```{.python .input  n=5}
+```{.python .input}
 net_G = nn.Sequential()
 net_G.add(nn.Dense(2))
 ```
 
+```{.python .input}
+#@tab pytorch
+net_G = nn.Sequential()
+net_G.add_module("Linear", nn.Linear(2, 2))
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/gan.md:120]:
```suggestion
net_G = nn.Sequential(nn.Linear(2, 2))
```
We are defining all the layers at once in Sequential for consistency.

File: chapter_generative-adversarial-networks/gan.md [hunk 3]
```diff
@@ -42,65 +42,112 @@ $$min_D max_G \{ -E_{x \sim \text{Data}} log D(\mathbf x) - E_{z \sim \text{Nois
...
 
+```{.python .input}
+#@tab pytorch
+net_D = nn.Sequential()
+net_D.add_module("Linear1t", nn.Linear(2, 5))
+net_D.add_module("Tanh1", nn.Tanh())
+net_D.add_module("Linear2t", nn.Linear(5, 3))
+net_D.add_module("Tanh2", nn.Tanh())
+net_D.add_module("Linear3", nn.Linear(3, 1))
+net_D.add_module("Sigmoid", nn.Sigmoid())
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/gan.md:142]: We are defining all the layers at once in Sequential for consistency.

File: chapter_generative-adversarial-networks/gan.md [hunk 4]
```diff
@@ -135,9 +201,26 @@ def update_G(Z, net_D, net_G, loss, trainer_G):  #@save
...
+    ones = torch.ones(batch_size)
+    trainer_G.zero_grad()
+    # We could reuse `fake_X` from `update_D` to save computation
+    fake_X = net_G(Z)
+    # Recomputing `fake_Y` is needed since `net_D` is changed
+    fake_Y = net_D(fake_X)
+    loss_G=loss(fake_Y,ones)
+    loss_G.backward()
+    trainer_G.step()
+    return float(loss_G.sum())
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/gan.md:218]:
```suggestion
    return loss_G
```
nn.BCELoss already uses mean reduction

File: chapter_generative-adversarial-networks/gan.md (hunk 3 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/gan.md:142]: I think `add_module` is more simliar to `add` in mxnet.

File: chapter_generative-adversarial-networks/gan.md (hunk 2 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/gan.md:120]: I think add_module is more simliar to add in mxnet.

File: chapter_generative-adversarial-networks/gan.md (hunk 1 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/gan.md:59]: `d2lbook build lib` doesn't appear in contribution.md. I think we should add it.

File: chapter_generative-adversarial-networks/gan.md (hunk 2 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/gan.md:120]: Both way is ok.https://discuss.pytorch.org/t/when-to-use-add-module-function/10534.

File: chapter_generative-adversarial-networks/gan.md (hunk 3 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/gan.md:142]: ok.. I'll change it just for pytorch's consistency, not for silimarity.

**Comment by StevenJokes** [comment]
image
https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/PR1308.ipynb
Why 10 times there?

**Comment by StevenJokes** [comment]
https://github.com/d2l-ai/d2l-en/issues/1300
